src/na/na_ucx.c

```diff
@@ -860,6 +881,45 @@ na_ucx_initialize(na_class_t *nacl, const struct na_info NA_UNUSED *na_info,
     }
 #endif
 
+    if (na_info->na_init_info != NULL) {
+        /* Thread mode */
+        if ((na_info->na_init_info->max_contexts > 1)
+            && !(na_info->na_init_info->thread_mode & NA_THREAD_MODE_SINGLE)) {
+            /* If the UCP context can potentially be used by more than one
+             * worker / thread, then this context needs thread safety. */
+            global_params.field_mask |= UCP_PARAM_FIELD_MT_WORKERS_SHARED;
+            global_params.mt_workers_shared = 1;
+            context_thread_mode = UCS_THREAD_MODE_MULTI;
+        }
+        if (na_info->na_init_info->thread_mode & NA_THREAD_MODE_SINGLE_CTX)
+            worker_thread_mode = UCS_THREAD_MODE_SINGLE;
+    }
+
+#ifdef NA_UCX_HAS_LIB_QUERY
+    ucp_lib_attrs.field_mask = UCP_LIB_ATTR_FIELD_MAX_THREAD_LEVEL;
+    status = ucp_lib_query(&ucp_lib_attrs);
+
+    if (status != UCS_OK) {
+        NA_LOG_ERROR("ucp_context_query: %s", ucs_status_string(status));
+        ret = NA_PROTOCOL_ERROR;
+        goto cleanup;
+    }
+
+    if ((ucp_lib_attrs.field_mask & UCP_LIB_ATTR_FIELD_MAX_THREAD_LEVEL) == 0) {
+        NA_LOG_ERROR("lib attributes contain no max thread level");
+        ret = NA_PROTOCOL_ERROR;
+        goto cleanup;
+    }
+
+    /* Best effort to ensure thread safety */
+    if (worker_thread_mode != UCS_THREAD_MODE_SINGLE
+        && ucp_lib_attrs.max_thread_level < worker_thread_mode) {
+        worker_thread_mode = ucp_lib_attrs.max_thread_level;
+        NA_LOG_WARNING("Max worker thread level is: %s",
+            ucs_thread_mode_names[worker_thread_mode]);
+    }
+#endif
```

Comment:
The plugin should honor the thread mode that the caller requests or else fail with a meaningful error return.

Reply:
ok yes was trying to be a little permissive but probably better to fail, will fix